grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy

```diff
@@ -496,4 +532,195 @@ class StockMovementApiController {
+
+        def shipmentItems = []
+        shipments.each { Shipment shipment ->
+            shipment.shipmentItems.findAll { it.quantityRemaining > 0 }.groupBy {
+                it.product
+            }.each { product, value ->
+                shipmentItems << [
+                    productCode         : product.productCode,
+                    productName         : product.name,
+                    quantity            : value.sum { it.quantityRemaining },
+                    expectedShippingDate: formatDate(date: shipment.expectedShippingDate, format: "dd-MMM-yy"),
+                    expectedDeliveryDate: formatDate(date: shipment.expectedDeliveryDate, format: "dd-MMM-yy"),
+                    shipmentNumber      : shipment.shipmentNumber,
+                    shipmentName        : shipment.name,
+                    origin              : shipment.origin,
+                    destination         : shipment.destination,
+                ]
+            }
+        }
+
+        if (shipmentItems) {
+            def csv = getShipmentItemsCsv(shipmentItems)
+            def date = new Date()
+            response.contentType = "text/csv"
+            response.setHeader("Content-disposition", "attachment; filename=\"Items shipped not received_${destination.name}_${date.format("yyyyMMdd-hhmmss")}.csv\"")
+            render(contentType: "text/csv", text: csv.out.toString())
+            return
+        } else {
+            def message = "No shipment items found"
+            response.status = 404
+            render([errorMessage: message] as JSON)
+            return
+        }
+    }
+
+    /**
+     * Action to get pending requisition items as csv
+     * */
+    def pendingRequisitionItems = {
+        if (!params.origin) {
+            def message = "Origin parameter cannot be the empty"
+            response.status = 400
+            render([errorMessages: [message]] as JSON)
+            return
+        }
+
+        def origin = Location.get(params.origin)
+
+        if (!origin) {
+            def message = "Origin does not exist"
+            response.status = 404
+            render([errorMessage: message] as JSON)
+            return
+        }
+
+        def pendingRequisitionItems = stockMovementService.getPendingRequisitionItems(origin)
+
+        if (pendingRequisitionItems) {
+            def csv = getPendingRequisitionItemsCsv(pendingRequisitionItems)
+            def date = new Date()
+            response.contentType = "text/csv"
+            response.setHeader("Content-disposition", "attachment; filename=\"PendingShipmentItems-${new Date().format("yyyyMMdd-hhmmss")}.csv\"")
+            render(contentType: "text/csv", text: csv.out.toString())
+            return
+        } else {
+            def message = "No pending requisition items found"
+            response.status = 404
+            render([errorMessage: message] as JSON)
+            return
+        }
+    }
+
+    def getShipmentItemsCsv(List shipmentItems) {
```

Comment:
In general, I don't like these methods here in the controller. They're better off in a service. I also think all of these could be more generalizable. We could convert each of these objects into a map and pass the map to one renderCsv method that does something very simple like this.

```
def keys = data[0].keySet() 
csv.printRecord(keys)
data.each { row ->
    csv.printRecord(row.values)
}
```

Reply:
And then all you need to do is convert the shipment or requisition item or whatever into a data map.
```
def data = shipments.collect { 
   return [
            "Code" : shipmentItem.productCode,
            "Product Name": shipmentItem.productName,
            "Quantity Incoming": shipmentItem.quantity,
            "Expected Shipping Date": shipmentItem.expectedShippingDate,
            "Expected Delivery Date": shipmentItem.expectedDeliveryDate,
            "Shipment Number": shipmentItem.shipmentNumber,
            "Shipment Name": shipmentItem.shipmentName,
            "Origin": shipmentItem.origin,
            "Destination": shipmentItem.destination,
  ]
}
```